Google.Api.Gax.Grpc/CallSettings.cs

```diff
@@ -309,19 +332,37 @@ internal static void Concatenate(Metadata entries, string name, string separator
                 GaxPreconditions.CheckNotNull(name, nameof(name));
                 GaxPreconditions.CheckNotNull(separator, nameof(separator));
 
-                var currentValues = entries.GetAll(name).Select(entry => entry.Value);
-                if (currentValues.Skip(1).Any()) // Find out if we have 2 or more elements, efficiently.
+                // Find the first entry to concatenate.
+                int firstIndex;
+                for (firstIndex = 0; firstIndex < entries.Count && entries[firstIndex].Key != name; firstIndex++);
+
+                if (firstIndex == entries.Count) // There are no values associated to name.
                 {
-                    var newValue = string.Join(separator, currentValues);
-                    // Note that string.Join iterates the whole entry collection,
-                    // and calling Override will do the same to remove the entries.
-                    // We could concatenate and remove each entry on a single iteration
-                    // but if there's an error within the iterations, for instance, because
-                    // an entry has a bytes values instead of a string value, then the entry
-                    // set will remain in an incosistent state, where some of the existing
-                    // entries have been removed but not all, and the concatenated value has not been added.
-                    Override(entries, name, newValue);
+                    return;
                 }
+
+                string firstValue = entries[firstIndex].Value;
+                StringBuilder builder = new StringBuilder(firstValue);
```

Comment:
Possibly keep it null to start with, and initialize it when we see the second entry?
(This is possibly overkill, but it seems like a simple enough change, to avoid allocating in the common case of there only being one entry.)